crates/bevy_asset/src/asset_changed.rs

```diff
@@ -0,0 +1,446 @@
+            check: self.check,
+        }
+    }
+}
+
+/// State for [`AssetChanged`].
+#[doc(hidden)]
+pub struct AssetChangedState<A: AsAssetId> {
+    asset_id: ComponentId,
+    resource_id: ComponentId,
+    _asset: PhantomData<fn(A)>,
+}
+
+// SAFETY: `ROQueryFetch<Self>` is the same as `QueryFetch<Self>`
+unsafe impl<A: AsAssetId> WorldQuery for AssetChanged<A> {
+    type Item<'w> = ();
+    type Fetch<'w> = AssetChangedFetch<'w, A>;
+
+    type State = AssetChangedState<A>;
+
+    fn shrink<'wlong: 'wshort, 'wshort>(_: QueryItem<'wlong, Self>) -> QueryItem<'wshort, Self> {}
+
+    fn shrink_fetch<'wlong: 'wshort, 'wshort>(fetch: Self::Fetch<'wlong>) -> Self::Fetch<'wshort> {
+        fetch
+    }
+
+    unsafe fn init_fetch<'w>(
+        world: UnsafeWorldCell<'w>,
+        state: &Self::State,
+        last_run: Tick,
+        this_run: Tick,
+    ) -> Self::Fetch<'w> {
+        let err_msg = || {
+            panic!(
+                "AssetChanges<{ty}> resource was removed, please do not remove \
+                AssetChanges<{ty}> when using the AssetChanged<{ty}> world query",
+                ty = ShortName::of::<A>()
+            )
+        };
+        // SAFETY: `AssetChanges` is private and only accessed mutably in the `AssetEvents` schedule
+        let changes: &AssetChanges<_> = unsafe { world.get_resource().unwrap_or_else(err_msg) };
+        let has_updates = changes.last_change_tick.is_newer_than(last_run, this_run);
+
+        AssetChangedFetch {
+            inner: has_updates.then(||
+                    // SAFETY: We delegate to the inner `init_fetch` for `A`
+                    unsafe {
+                        <&A>::init_fetch(world, &state.asset_id, last_run, this_run)
+                    }),
+            check: AssetChangeCheck::new(changes, last_run, this_run),
+        }
+    }
+
+    const IS_DENSE: bool = <&A>::IS_DENSE;
+
+    unsafe fn set_archetype<'w>(
+        fetch: &mut Self::Fetch<'w>,
+        state: &Self::State,
+        archetype: &'w Archetype,
+        table: &'w Table,
+    ) {
+        if let Some(inner) = &mut fetch.inner {
+            // SAFETY: We delegate to the inner `set_archetype` for `A`
+            unsafe {
+                <&A>::set_archetype(inner, &state.asset_id, archetype, table);
+            }
+        }
+    }
+
+    unsafe fn set_table<'w>(fetch: &mut Self::Fetch<'w>, state: &Self::State, table: &'w Table) {
+        if let Some(inner) = &mut fetch.inner {
+            // SAFETY: We delegate to the inner `set_table` for `A`
+            unsafe {
+                <&A>::set_table(inner, &state.asset_id, table);
+            }
+        }
+    }
+
+    unsafe fn fetch<'w>(_: &mut Self::Fetch<'w>, _: Entity, _: TableRow) -> Self::Item<'w> {}
+
+    #[inline]
+    fn update_component_access(state: &Self::State, access: &mut FilteredAccess<ComponentId>) {
+        <&A>::update_component_access(&state.asset_id, access);
+        access.add_resource_read(state.resource_id);
```

Comment:
I don't think this line is quite enough.  We need `Query` to report the access in `Access<ArchetypeComponentId>` for it to be used by the scheduler, and it currently does that in `QueryState::update_archetype_component_access()`, which only looks at `component_reads_and_writes()` and `component_writes()`.  

I think you need modify `QueryState::new_with_access()` to look at `resource_reads_and_writes()` and `resource_writes()` and convert them to `ArchetypeComponentId`s.  

I think you also need to modify `FilteredAccess::is_compatible()` to consider resources to conflict even when the filters don't overlap.  (That may need to get reworked in the future for resources-as-components, but the semantics seem reasonable.)  

It would be good to add a unit test that verifies that the access is set up correctly.  Maybe check that `fn system(r: ResMut<AssetChanges<A>>, q: Query<Entity, AssetChanged<A>>)` panics when passed to `assert_system_does_not_conflict`, and again with the parameters reversed.

Reply:
Alright thanks for the close review. Will add a test to validate.

Reply:
So just added a test and without making any other changes am getting the following error:
```rust
error[B0001]: Query<MyComponent, AssetChanged<MyComponent>> in system bevy_asset::asset_changed::tests::resource_conflict::system accesses component(s)AssetChanges<MyAsset> in a way that conflicts with a previous system parameter. Consider using `Without<T>` to create disjoint Queries or merging conflicting Queries into a `ParamSet`. See: https://bevyengine.org/learn/errors/b0001
```

Thoughts?

Reply:
Oh, right, `ResMut` does unfiltered access, so `FilteredAccess::is_compatible()` will see that the filters overlap and catch it.  Great!  So that won't be a problem unless someone tries to do *mutable* resource access from a query.  

I think it's just the `Access<ArchetypeComponentId>` that would be an issue, then.  I'm not coming up with a simple way to write a test for that, though.

Reply:
So, I think it's just adding the following to `QueryState::update_archetype_component_access`? https://github.com/bevyengine/bevy/pull/16810/commits/d0130a34108ece47c77232d3b4a6e85874090547

Reply:
Not quite.  That adds access to components on entities in archetypes matched by the query.  But you want to add access to resources from the world.  Resources on the world have their own `ArchetypeComponentId`s that don't appear in any archetypes.  

If we just want to support this use case of a finite set of read access, I think it's enough to add code to `Query::new_with_access` that does
```rust
for component_id in state.component_access.resource_reads() {
    access.add_resource_read(world.initialize_resource_internal(component_id).id());
}
```

More complete support would need to check for writes and for `reads_all` and `writes_all`, and look a bit like the code in `FilteredResourcesMutParamBuilder` https://github.com/bevyengine/bevy/blob/6178ce93e8537f823685e8c1b617e22ba93696e7/crates/bevy_ecs/src/system/builder.rs#L644

---

The background here is that adding access in a `QueryData` has only ever added access to entities matched by the query.  There has never been a way to declare that you want access to resources from the world.  

But the code you wrote *looks* like it should give access to resources!  And that access is useful!  So I think we should add the plumbing to make it work.  

It might be cleaner to do that in a separate PR, though, and put this one back to what you had originally.  This one would just be unsound until the other one was merged.  If you'd like, I can try to put one together on Monday.

Reply:
Thanks @chescock, that would be super helpful. I'm mostly trying to land this to unblock some rendering features that depend on asset change detection and some of the ECS internals are a bit above my paygrade. Appreciate your education here, as it's better building my understanding.

Reply:
> But the code you wrote looks like it should give access to resources! And that access is useful! So I think we should add the plumbing to make it work.

Yes please! This is really useful for unlocking indexes as a first class ECS feature.

Reply:
Okay!  #16843

While writing tests there, I also noticed we might need to add
```rust
assert!(!access.access().has_resource_write(component_id), "some useful message");
```
just before the `add_resource_read` call, unless we decide to just ban queries from making mutable resource access.

Reply:
I've removed the changes I added and merged your changes from main. I've also added the assertion for write access in the filter.